### PR TITLE
Make tf.errors subclasses of relevant Python builtin exceptions

### PR DESCRIPTION
**Pull request by lgeiger**
Whe porting some code from Python's builtin file I/O to `tf.io.gfile` I had to adapt a lot of error handling which was checking against the [builtin Python exceptions](https://docs.python.org/3.8/library/exceptions.html#exception-hierarchy).

This PR makes the following `tf.errors` subclasses of their respective builtin Python analogue so that users don't need to modify `isinstance` checks during error handling when switching between Python's file I/O and `tf.io.gfile`:
- `tf.errors.NotFoundError` subclasses [`FileNotFoundError`](https://docs.python.org/3.8/library/exceptions.html#FileNotFoundError)
- `tf.errors.AlreadyExistsError` subclasses [`FileExistsError`](https://docs.python.org/3.8/library/exceptions.html#FileExistsError)
- `tf.errors.PermissionDeniedError` subclasses [`PermissionError`](https://docs.python.org/3.8/library/exceptions.html#PermissionError)
- `tf.errors.UnimplementedError` subclasses [`NotImplementedError`](https://docs.python.org/3.8/library/exceptions.html#NotImplementedError)

Together with #41173 this should improve the usability of `tf.io.gfile` as a replacement for Python's file I/O.

One could also think of making [`tf.errors.OutOfRangeError`](https://www.tensorflow.org/api_docs/python/tf/errors/OutOfRangeError) a subclass of [`StopIteration`](https://docs.python.org/3.8/library/exceptions.html#StopIteration) but I am not sure if there are some subtle differences between the meaning of those two errors.

### CODE REVIEW

**Review by mihaimaruseac — DISMISSED** [review]

**Review by av8ramit — DISMISSED** [review]

**Comment by lgeiger** [comment]
I had to update the API goldens since the change in inheritance would fail CI otherwise.

**Review by mihaimaruseac — APPROVED** [review]

File: tensorflow/python/framework/errors_impl.py [hunk 1]
```diff
@@ -283,7 +283,7 @@ def __init__(self, node_def, op, message):
 
 
 @tf_export("errors.NotFoundError")
-class NotFoundError(OpError):
+class NotFoundError(OpError, FileNotFoundError):
```
Comment (mihaimaruseac) [inline tensorflow/python/framework/errors_impl.py:286]: Hmm, we are returning `NotFoundError` in other cases too. See for example https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/framework/op_kernel.cc;l=1417;drc=a565c473c14da0c1bb35bd1771bdc195a72f7982

File: tensorflow/python/framework/errors_impl.py [hunk 2]
```diff
@@ -300,7 +300,7 @@ def __init__(self, node_def, op, message):
 
 
 @tf_export("errors.AlreadyExistsError")
-class AlreadyExistsError(OpError):
+class AlreadyExistsError(OpError, FileExistsError):
```
Comment (mihaimaruseac) [inline tensorflow/python/framework/errors_impl.py:303]: Also used with other semantics

https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/kernels/batch_kernels.cc;l=1299;drc=fbac6b974de07a175c8b8d65912043b2f4bd1bc3

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]
Also, can you add a short change description to `RELEASE.md` to document the API change?

**Comment by mihaimaruseac** [comment]
We need to not subclass the two exceptions that are used outside of file scope due to the fact that we use the TF ones in more places

**Comment by lgeiger** [comment]
> We need to not subclass the two exceptions that are used outside of file scope due to the fact that we use the TF ones in more places

Hmm in that case I am not sure if this PR is actually an overall improvement (the motivation for this PR was mainly the `AlreadyExistsError` and `NotFoundError` which are mainly used as filesystem errors in TF).
I would argue that having some errors subclass the related builtin exceptions and some errors not, is more confusing to users than not subclassing any of them.

**Comment by mihaimaruseac** [comment]
Yes, I think it will be more confusing to do so.

Unfortunately our C++ exception hierarchy does not fit 100% to the standard Python one and it would be a lot of work to migrate.

**Comment by gbaned** [comment]
@lgeiger  Can you please resolve conflicts? Thanks!

**Review by mihaimaruseac — APPROVED** [review]

**Comment by gbaned** [comment]
@lgeiger Can you please resolve conflicts? Thanks!

**Comment by mihaimaruseac** [comment]
It seems there are thousands of internal tests failing. Looking into it, probably some import is missing. If that is not the case, I'll be back here

**Comment by lgeiger** [comment]
> Are these errors different in py2?

It looks like `PermissionError` doesn't exist in Python 2.

> The failing test are caused by running internal tests with Python2, since internally we still need to support some of these while they move forward.

Is there a timeline for when Python 3 only code can be submitted to the TensorFlow code base?

Since we cannot really change `NotFoundError` and `AlreadyExistsError` anyway, I won't spend time making this PR Python 2 compatible. It also looks like it is accumulating a lot of merge conflicts in the release notes.
I am happy to resubmit once TensorFlow is Python 3 only.

**Comment by mihaimaruseac** [comment]
Apologies for the wasted time/effort.

Unfortunately, we don't have a public guarantee on when Python 2 support can be fully dropped. External PRs must only be Python3 and code to TF must be only Python3 even if internally. But if an internal test that uses Py2 breaks we either need to wait for that test to fix in sync with the change or delay/reject the change. Sorry for all the issues this causes

**Comment by lgeiger** [comment]
> Apologies for the wasted time/effort.

That's alright! @mihaimaruseac you are very responsive on GitHub so not much effort was wasted here. Not all PRs can be accepted that's normal (even if it's due to internal reasons).
I have other open PRs where where I'm waiting a few months to get a review, that's much more frustrating, than a PR that doesn't get accepted :)

> Unfortunately, we don't have a public guarantee on when Python 2 support can be fully dropped. External PRs must only be Python3 and code to TF must be only Python3 even if internally. But if an internal test that uses Py2 breaks we either need to wait for that test to fix in sync with the change or delay/reject the change. Sorry for all the issues this causes

I see, thanks for the explanation.

**Comment by mihaimaruseac** [comment]
Apologies for the stalled PRs. Can you ping me on them to take a look, as possible?

Thank you

**Pull request closed without merging**